Данная работа посвящена шифрованию и расшифровке шифротекстов, закодированных простыми классическими шифрами.

In [1]:
%load_ext pycodestyle_magic
%pycodestyle_on #or %flake8_on
%flake8_on


Расшифровка текста
==================

Для хранения и передачи секретных данных требуется либо использовать защищённый канал связи, либо опеспечить шифрование самого сообщения, чтобы даже при его перехвате злоумышленник не смог его прочитать. Наука о шифровании данных называется __криптография__. (Существует ещё вариант сокрытия самого факта передачи информации — __стеганография__.)


Атбаш
-----

Атбаш — это шифр простой замены, использованный для [еврейского алфавита](https://ru.wikipedia.org/wiki/%D0%95%D0%B2%D1%80%D0%B5%D0%B9%D1%81%D0%BA%D0%BE%D0%B5_%D0%BF%D0%B8%D1%81%D1%8C%D0%BC%D0%BE) и получивший оттуда своё название. Шифрование происходит заменой первой буквы алфавита на последнюю, второй на предпоследнюю и так далее.

Для русского алфавита таблица замен:


|а|б|в|г|д|е|ё|ж|з|и|й|к|л|м|н|о|п|р|с|т|у|ф|х|ц|ч|ш|щ|ъ|ы|ь|э|ю|я|

|я|ю|э|ь|ы|ъ|щ|ш|ч|ц|х|ф|у|т|с|р|п|о|н|м|л|к|й|и|з|ж|ё|е|д|г|в|б|а|


Данный шифр тривиален для взлома, поскольку у него нет ключа. Если криптоаналитику известно, что текст зашифрован этим шифром, то он сразу применяет алгоритм дешифрования, эквивалентный шифрованию.

Программа для шифрования шифром "атбаш":

```python

	class Atbash:
	    alphabet = "абвгдеёжзийклмнопрстуфхцчшщъыьэюя"

	    def __init__(self):
	        lowercase_code = {x: y for x, y in zip(self.alphabet, self.alphabet[::-1])}
	        uppercase_code = {x.upper(): y.upper() for x, y in zip(self.alphabet, self.alphabet[::-1])}
	        self._encode = lowercase_code
	        self._encode.update(uppercase_code)

	    def encode(self, text):
	        return ''.join([self._encode.get(char, char) for char in text])


	cipher = Atbash()
	line = input()
	while line != '.':
	    print(cipher.encode(line))
	    line = input()
```

Следующий раздел лабораторной работы закодирован шифром "атбаш".

In [7]:
class Atbash:
    alphabet = "абвгдеёжзийклмнопрстуфхцчшщъыьэюя"

    def __init__(self):
        lowercase_code = \
            {x: y for x, y in zip(self.alphabet, self.alphabet[::-1])}
        uppercase_code = {
            x.upper(): y.upper() for x, y in zip(
                self.alphabet,
                self.alphabet[::-1]
            )
        }
        self._encode = lowercase_code
        self._encode.update(uppercase_code)

    def encode(self, text):
        return ''.join([self._encode.get(char, char) for char in text])

    def decode(self, text):
        return ''.join([self._encode.get(char, char) for char in text])


cipher = Atbash()
line = input()
while line != '.':
    print(cipher.decode(line))
    line = input()


 s
 
 .


Жцко Иъчяоа
-----------
Шифр Цезаря
-----------
Шифр Цезаря — это вид шифра подстановки, в котором каждый символ в открытом тексте заменяется символом, находящимся на некотором постоянном числе позиций левее или правее него в алфавите. Например, в шифре со сдвигом вправо на 3, А была бы заменена на Г, Б станет Д, и так далее.  

Шифр назван в честь римского императора Гая Юлия Цезаря, использовавшего его для секретной переписки со своими генералами.  

Следующая часть работы зашифрована шифром Цезаря.


```python

	class Caesar:
	    alphabet = "яюэьыъщшчцхфутсрпонмлкйизжёедгвба"

	    def __init__(self, key):
	        self._encode = dict()
	        for i in range(len(self.alphabet)):
	            letter = self.alphabet[i]
	            encoded = self.alphabet[(i + key) % len(self.alphabet)]
	            self._encode[letter] = encoded
	            self._encode[letter.upper()] = encoded.upper()
	        self._decode = {}  # TODO

	    def encode(self, text):
	        return ''.join([self._encode.get(char, char) for char in text])

	    def decode(self, line):
	    	pass  # TODO


	key = int(input('Ээъыцмъ фубз:'))
	cipher = Caesar(key)
	line = input()
	while line:
	    print(cipher.encode(line))
	    line = input()
```

Допишите метод decode и расшифруйте следующий раздел лабораторной работы. 
Подумайте, почему вам не сообщили ключ шифрования и что вам с этим делать.

In [26]:
class Caesar:
    alphabet = "яюэьыъщшчцхфутсрпонмлкйизжёедгвба"

    def __init__(self, key):
        self._encode = dict()
        for i in range(len(self.alphabet)):
            letter = self.alphabet[i]
            encoded = self.alphabet[(i + key) % len(self.alphabet)]
            self._encode[letter] = encoded
            self._encode[letter.upper()] = encoded.upper()
        self._decode = dict()
        for i in range(len(self.alphabet)):
            letter = self.alphabet[i]
            encoded = self.alphabet[(i - key) % len(self.alphabet)]
            self._decode[letter] = encoded
            self._decode[letter.upper()] = encoded.upper()

    def encode(self, text):
        return ''.join([self._encode.get(char, char) for char in text])

    def decode(self, text):
        return ''.join([self._decode.get(char, char) for char in text])


key = int(input('Ээъыцмъ фубз:'))
cipher = Caesar(key)
line = input()
while line:
    print(cipher.decode(line))
    line = input()


Ээъыцмъ фубз: 9
 .
 


Шифр простой замены
-------------------

Поздравляем с расшифровкой раздела!

Итак, вы догадались почему шифр Цезаря не является криптостойким: слишком мала мощность множества ключей и нужный ключ легко найти методом полного перебора.

Можно ли увеличить криптостойкость, не меняя метод шифрования? Да, можно. Если заменять один символ алфавита на определённый другой символ того же алфавита по какой-то таблице замен, то сама таблица замен и является ключом.

|а|б|в|г|д|е|ё|ж|з|и|й|к|л|м|н|о|п|р|с|т|у|ф|х|ц|ч|ш|щ|ъ|ы|ь|э|ю|я|

|ь|з|ц|в|к|б|щ|х|ф|п|р|я|л|э|а|е|ы|ш|у|м|ъ|ж|ё|о|г|й|н|д|и|т|ч|ю|с|


Множество ключей — это множество возможных таблиц простых замен.

Для русского алфавита мощность множества таблиц простых замен равна факториалу от 33.

33! = 8683317618811886495518194401280000000

Если тратить на проверку одного варианта 0.000001 секунды, получится 2.8e+23 лет.

Может показаться, что шифр простой замены вполне криптостойкий, однако это не так.

Его достаточно просто взломать при помощи [йтдебеабхб татюыът](https://ru.wikipedia.org/wiki/%D0%A7%D0%B0%D1%81%D1%82%D0%BE%D1%82%D0%BD%D1%8B%D0%B9_%D0%B0%D0%BD%D0%B0%D0%BB%D0%B8%D0%B7). Дело в том, что частота появления заданной буквы алфавита в достаточно длинных текстах одна и та же для разных текстов одного языка.

Если в шифротексте будет символ с вероятностью появления, аналогичной стандартной для языка, то можно предположить, что он и является указанной зашифрованной буквой.


Метод частотного криптоанализа известен с IX-го века, хотя наиболее известным случаем его применения в реальной жизни, возможно, является дешифровка египетских иероглифов в 1822 году.


Итак, следующая часть работы зашифрована при помощи следующей программы:

```python
	import random

	class Monoalphabet:
	    alphabet = ""  # TODO

	    def __init__(self, keytable):
	        lowercase_code = {x: y for x, y in zip(self.alphabet, keytable)}
	        uppercase_code = {x.upper(): y.upper() for x, y in zip(self.alphabet, keytable)}
	        self._encode = lowercase_code
	        self._encode.update(uppercase_code)
	        self._decode = {}  # TODO

	    def encode(self, text):
	        return ''.join([self._encode.get(char, char) for char in text])

	    def decode(self, line):
	        pass  # TODO


	key = Monoalphabet.alphabet[:]
	random.shuffle(key)
	cipher = Monoalphabet(key)
	line = input()
	while line:
	    print(cipher.encode(line))
	    line = input()
```

Программу для частотного анализа следует написать самостоятельно. Успехов!

In [6]:
class Monoalphabet:
    alphabet = "ЕГПЖОНЩЮЭБЭЧМЯХЫЦИЛДШАРСЙЗВЬЁКТФ"

    def __init__(self, keytable):
        lowercase_code = {x: y for x, y in zip(self.alphabet, keytable)}
        uppercase_code = {
            x.lower(): y.lower() for x,
            y in zip(self.alphabet, keytable)
        }
        self._encode = lowercase_code
        self._encode.update(uppercase_code)
        self._decode = {}
        lowercase_code = {
            x: y for y,
            x in zip(self.alphabet, keytable)
        }
        uppercase_code = {
            x.lower(): y.lower() for y, x in zip(self.alphabet, keytable)
        }
        self._decode = lowercase_code
        self._decode.update(uppercase_code)

    def encode(self, text):
        return ''.join([self._encode.get(char, char) for char in text])

    def decode(self, text):
        return ''.join([self._decode.get(char, char) for char in text])


key = 'ШИФРВЖЕНАТАКДЛЬЙОЯЧЁСПГМЗЫХБЦЭУЮ'
cipher = Monoalphabet(key)
line = input()
while line:
    print(cipher.encode(line))
    line = input()


 j
 


36:1: W391 blank line at end of file


In [4]:
a = str(input())
a = a.lower()
a = []
for i in "абвгдеёжзийклмнопрстуфхцчшщъыьэюя":
    if i in a:
        w = [a.count(i), i]
    else:
        w = [0, i]
    a.append(w)
a.sort(key=lambda x: x[0])
key = []
pre = 'оеаинтсрвлкмдпуяыьгзбчйхжшюцщэфъё'
di = dict()
for i in range(len(a)):
    di[pre[i]] = a[i][1]
print(di)


 Дальнейшим продолжением шифров простой замены является многоалфавитные шифры. Абу Аль-Кинди в своих работах показал, что обычные моноалфавитные шифры довольно-таки просто поддаются частотному криптоанализу и первым предложил использовать многоалфавитные шифры. В Европе такие шифры были впервые описаны в 1467 году итальянским архитектором Леон Баттиста Альберти. В XVI веке немецкий аббат Иоганн Тритемий в своей книге “Стенография” представил схему полиалфавитного шифрования в виде таблицы. Более сложный вариант с использованием смешанных алфавитов был описан в 1563 году Джамбаттиста делла Порта в его книге “Про скрытую значимость отдельных букв”. Последним словом в развитии полиалфавитных шифров можно считать роторные машины, примером которой можно считать немецкую машину Enigma, разработанная в 1917 г. Суть полиалфавитных шифров заключена в многократном применении различных шифров простой замены к определенному числу букв шифруемого текста. То есть к каждой букве по отдельности примен

ValueError: too many values to unpack (expected 3)

{'о': 'а', 'е': 'б', 'а': 'в', 'и': 'г', 'н': 'д', 'т': 'е', 'с': 'ё', 'р': 'ж', 'в': 'з', 'л': 'и', 'к': 'й', 'м': 'к', 'д': 'л', 'п': 'м', 'у': 'н', 'я': 'о', 'ы': 'п', 'ь': 'р', 'г': 'с', 'з': 'т', 'б': 'у', 'ч': 'ф', 'й': 'х', 'х': 'ц', 'ж': 'ч', 'ш': 'ш', 'ю': 'щ', 'ц': 'ъ', 'щ': 'ы', 'э': 'ь', 'ф': 'э', 'ъ': 'ю', 'ё': 'я'}


AttributeError: '_io.StringIO' object has no attribute 'buffer'

In [50]:
s = list(map(str, input().split()))
alf = ''
key = ''
for i in s:
    alf += i[0]
    key += i[1]
print(alf)
print(key)


 ЕШ ГИ ПФ ЖР ОВ НЖ ЩЕ ЮН ЭА БТ ЭА ЧК МД ЯЛ ХЬ ЫЙ ЦО ИЯ ЛЧ ДЁ ШС АП РГ СМ ЙЗ ЗЫ ВХ ЬБ ЁЦ КЭ ТУ ФЮ


ЕГПЖОНЩЮЭБЭЧМЯХЫЦИЛДШАРСЙЗВЬЁКТФ
ШИФРВЖЕНАТАКДЛЬЙОЯЧЁСПГМЗЫХБЦЭУЮ


Шифр Виженера (полиалфавитной замены)
-------------------------------------

Дальнейшим продолжением шифров простой замены является многоалфавитные шифры. Абу Аль-Кинди в своих работах показал, что обычные моноалфавитные шифры довольно-таки просто поддаются частотному криптоанализу и первым предложил использовать многоалфавитные шифры. В Европе такие шифры были впервые описаны в 1467 году итальянским архитектором Леон Баттиста Альберти. В XVI веке немецкий аббат Иоганн Тритемий в своей книге “Стенография” представил схему полиалфавитного шифрования в виде таблицы. Более сложный вариант с использованием смешанных алфавитов был описан в 1563 году Джамбаттиста делла Порта в его книге “Про скрытую значимость отдельных букв”. Последним словом в развитии полиалфавитных шифров можно считать роторные машины, примером которой можно считать немецкую машину Enigma, разработанная в 1917 г. Суть полиалфавитных шифров заключена в многократном применении различных шифров простой замены к определенному числу букв шифруемого текста. То есть к каждой букве по отдельности применяется один из шифров простой замены.

Шифр Виженера состоит из последовательности нескольких шифров Цезаря с различными значениями сдвига. Для зашифровывания может использоваться таблица алфавитов, называемая квадрат (таблица) Виженера. Применительно к русскому алфавиту таблица Виженера составляется из строк по 33 символов, причём каждая следуюъая строка сдвигается на несколько позиций. Таким образом, в таблице получается 33 различных шифров Цезаря. На разных этапах кодировки шифр Виженера использует различные алфавиты из этой таблицы. На каждом этапе шифрования используются различные алфавиты, выбираемые в зависимости от символа ключевого слова. Например, если ключевое слово “САТ”, то первая буква открытого текста шифруется с использованием алфавита “С’, вторая “А”, третья “Т”, четвёртая снова “С” и так далее.

Программа шифрования шифром Виженера:
```python
	class Vigenere:
	    alphabet = ""  # TODO

	    def __init__(self, keyword):
	        self.alphaindex = {ch: index for index, ch in enumerate(self.alphabet)}
	        self.key = [self.alphaindex[letter] for letter in keyword.lower()]

	    def caesar(self, letter, shift):
	        if letter in self.alphaindex:  # строчная буква
	            index = (self.alphaindex[letter] + shift)%len(self.alphabet)
	            cipherletter = self.alphabet[index]
	        elif letter.lower() in self.alphaindex:  # заглавная буква
	            cipherletter = self.caesar(letter.lower(), shift).upper()
	        else:
	            cipherletter = letter
	        return cipherletter

	    def encode(self, line):
	        ciphertext = []
	        for i, letter in enumerate(line):
	            shift = self.key[i % len(self.key)]
	            cipherletter = self.caesar(letter, shift)
	            ciphertext.append(cipherletter)

	        return ''.join(ciphertext)

	    def decode(self, line):
	        pass  # TODO


	keyword = input('keyword=')
	cipher = Vigenere(keyword)

	line = input()
	while line != '.':
	    print(cipher.decode(line))
	    line = input()
```
Последний раздел работы зашифрован шифром Виженера с неизвестным кодовым словом. Подсказка для самых стойких криптоаналитиков: длина кодового слова 8.

In [8]:
class Vigenere:
    alphabet = "абвгдеёжзийклмнопрстуфхцчшщъыьэюя"

    def __init__(self, keyword):
        self.alphaindex = {ch: index for index, ch in enumerate(self.alphabet)}
        self.key = [self.alphaindex[letter] for letter in keyword.lower()]

    def caesar(self, letter, shift):
        if letter in self.alphaindex:  # строчная буква
            index = (self.alphaindex[letter] + shift) % len(self.alphabet)
            cipherletter = self.alphabet[index]
        elif letter.lower() in self.alphaindex:  # заглавная буква
            cipherletter = self.caesar(letter.lower(), shift).upper()
        else:
            cipherletter = letter
        return cipherletter

    def encode(self, line):
        ciphertext = []
        for i, letter in enumerate(line):
            shift = self.key[i % len(self.key)]
            cipherletter = self.caesar(letter, shift)
            ciphertext.append(cipherletter)

        return ''.join(ciphertext)

    def decode(self, line):
        text = []
        for i, letter in enumerate(line):
            shift = self.key[i % len(self.key)]
            if letter in self.alphaindex:
                index = (self.alphaindex[letter] - shift) % len(self.alphabet)
                text = self.alphabet[index]
            elif letter.lower() in self.alphaindex:
                text = self.decode(letter.lower())[0].upper()
            else:
                text = letter
            text.append(text)

        return ''.join(plaintext)


keyword = input('keyword=')
cipher = Vigenere(keyword)

line = input()
while line != '.':
    print(cipher.encode(line))
    line = input()

keyword= а
 фыф


фыф


 работает


работает


 но я не могу отгадать кодовое слово


но я не могу отгадать кодовое слово


 .


йвылщф
-------------------------------------
(ключ-слово: мфти)

ыцящгнюзрвхщдз щгхья дбахжтыи дгч эщтфхьтяхт дфыыачпг вчшэтфбффсявблы мыээф -- эчщдожящгцат ячрщюе [еэжщм "саспбн"](https://habr.com/ru/articles/721790/). ёшщянъжн хсжбмыц ксбебкмвыз, хёвчшръчоффбхйдз о бтбхвтю йжбт ющгсх, эдшыаоратеъл шб ъхй вчэ.